slides/aml-15-022818-resampling-imbalanced-data/index.html

```diff
@@ -0,0 +1,750 @@
+]
+
+???
+
+---
+
+# Basic Approaches
+
+.left-column[
+.center[
+![:scale 85%](images/basic_approaches.png)
+]
+]
+
+.right-column[
+</br>
+Change the training procedure
+]
+
+???
+
+---
+
+
+# Sckit-learn vs resampling
+
+- The transform method only transforms X
+- Pipelines work by chaining transforms
+- To resample the data, we need to also change y
+- Imbalance-learn extends scikit-learn interface with a
+“sample” method.
+- Imbalance-learn has a custom pipeline that allows
+resampling.
+- Imbalance-learn: resampling is only performed during fitting
+- Warning: not everything in imbalance-learn is multiclass!
+
+???
+
+---
+
+# Random Undersampling
+
+.smaller[
+- Drop data from the majority class randomly
+- Often untill balanced
+- Very fast training (data shrinks to 2x minority)
+- Loses data !
+```python
+from imblearn.under_sampling import RandomUnderSampler
+rus = RandomUnderSampler(replacement=False)
+X_train_subsample, y_train_subsample = rus.fit_sample(X_train, y_train)
+print(X_train.shape)
+print(X_train_subsample.shape)
+print(np.bincount(y_train_subsample))
+```
+(8387, 6) </br>
+(390, 6) </br>
+[195 195]
+]
+
+???
+
+
+---
+
+# Random Undersampling
+
+.smaller[
+```python
+from imblearn.pipeline import make_pipeline as make_imb_pipeline
+undersample_pipe = make_imb_pipeline(RandomUnderSampler(), LogisticRegressionCV())
+scores = cross_val_score(undersample_pipe, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.916512922589
+```python
+undersample_pipe_rf = make_imb_pipeline(RandomUnderSampler(), RandomForestClassifier())
+scores = cross_val_score(undersample_pipe_rf, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.942668403733
+
+- As accurate with fraction of samples!
+- Really good for large datasets
+]
+
+???
+
+---
+
+# Random Oversmapling
+
+.smaller[
+- Repeat samples from the minority class randomly
+- Often untill balanced
+- Much slower (dataset grows to 2x majority)
+```python
+from imblearn.over_sampling import RandomOverSampler
+ros = RandomOverSampler()
+X_train_oversample, y_train_oversample = ros.fit_sample(X_train, y_train)
+print(X_train.shape)
+print(X_train_oversample.shape)
+print(np.bincount(y_train_oversample))
+```
+(8387, 6) </br>
+(16384, 6) </br>
+[8192 8192]
+]
+???
+
+---
+
+# Random Oversampling
+
+.smaller[
+```python
+oversample_pipe = make_imb_pipeline(RandomOverSampler(), LogisticRegression())
+scores = cross_val_score(oversample_pipe, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.917755942193
+```python
+oversample_pipe_rf = make_imb_pipeline(RandomOverSampler(), RandomForestClassifier())
+scores = cross_val_score(oversample_pipe_rf, X_train, y_train, cv=10, scoring='roc_auc')
+print(np.mean(scores))
+```
+0.916313332777
+
+</br>
+</br>
+Logreg the same, Random Forest much worse than before
+]
+
+???
+
+---
+
+# ROC Curves for LogReg
+
+.center[
+![:scale 85%](images/roc_curves_for_logreg.png)
```

Comment:
can you redo the plot from the notebook and add a title so it's aligned with the plot on the next slide?